blog/2019-02-07-KIEServerOptaPlannerTaskAssignment.adoc

```diff
@@ -0,0 +1,215 @@
+- [Soft 2] Major tasks: Complete major tasks as soon as possible, sooner than minor tasks.
+- [Soft 3] Minor tasks: Complete minor tasks as soon as possible.
+
+=== Domain Model
+
+image::taskAssigningClassDiagram.png[link="taskAssigningClassDiagram.png"]
+You can read more on how to
+https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#modelAPlanningProblem[model a planning problem].
+
+=== https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#realTimePlanning[Real-time Planning]
+
+As new tasks come in and others start being implemented, the optimal solution might change. OptaPlanner handles such scenarios using
+https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#problemFactChange[ProblemFactChange].
+
+For example, if an employee starts working on a task, you can tell OptaPlanner to *pin* this task by sending a `PinTaskProblemFactChange`.
+Each `ProblemFactChange` type is implemented as a class that implements the interface `ProblemFactChange<Solution_>`.
+In the Task Assignment example, this class can be something like:
+[source,java]
+----
+public class PinTaskProblemFactChange implements ProblemFactChange<TaskAssigningSolution> {
+    ...
+
+    @Override
+    public void doChange(ScoreDirector<TaskAssigningSolution> scoreDirector) {
+        ...
+                scoreDirector.beforeProblemPropertyChanged(toBePinnedTask);
+                toBePinnedTask.setPinned(true);
+                scoreDirector.afterProblemPropertyChanged(toBePinnedTask);
+        ...
+
+        scoreDirector.triggerVariableListeners();
+    }
+}
+----
+The full class can be found https://github.com/kiegroup/kie-server-optaplanner-task-assignment/blob/master/task-assignment-kjar/src/main/java/org/kie/server/examples/optaplanner/taskassignment/kjar/domain/realtime/PinTaskProblemFactChange.java[here].
+
+===  https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#solverConfiguration[Solver Configuration]
+
+The solver configuration file determines how the solving process works. For a KJAR deployed to a KIE Server and since we are using
+Drools for score calculation, you need to use a `ksessionName`. This tells the KieContainer where to find the DRL file.
+Add this to `META-INF/kmodule.xml` file:
+[source,xml,options="nowrap"]
+----
+<kbase name="optataskKBase" packages="PATH_TO_SOLVER_RESOURCES">
+    <ksession name="optataskKsession"/>
+</kbase>
+----
+And to solver config file:
+[source,xml,options="nowrap"]
+----
+<scoreDirectorFactory>
+    <ksessionName>optataskKsession</ksessionName>
+</scoreDirectorFactory>
+----
+Notice that both solver configuration and DRL files need to be in the `resources` folder under the path
+`PATH_TO_SOLVER_RESOURCES`.
+
+For real-time planning, set the solver in
+https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#daemon[daemon mode] in order to resume
```

Comment:
replace "7.17.0.Final" by "latestFinal"
That latest is a symlink to the latest version:
  https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/

Ideally, we don't want website visitors reading old documentation (although we don't throw it away as they might be using an old version). This blog past can last for years. This does mean that links can become broken in the blog post if we change the anchors, but we rarely do that and replace them in the blog posts too if we do so.